qcodes_contrib_drivers/drivers/ERAInstruments/erasynth.py

```diff
@@ -0,0 +1,902 @@
+"""Driver for the ERASynth/ERASynth+/ERASynth++ signal generators.
+
+Author: Victor Negîrneac, vnegirneac@qblox.com
+
+For official instrument support visit:
+
+- https://erainstruments.com
+- https://github.com/erainstruments/
+
+This module provides the following drivers:
+
+- :class:`qcodes_contrib_drivers.drivers.ERAInstruments.ERASynth <qcodes_contrib_drivers.drivers.ERAInstruments.erasynth.ERASynth>`,
+- :class:`qcodes_contrib_drivers.drivers.ERAInstruments.ERASynthPlus <qcodes_contrib_drivers.drivers.ERAInstruments.erasynth.ERASynthPlus>`, and
+- :class:`qcodes_contrib_drivers.drivers.ERAInstruments.ERASynthPlusPlus <qcodes_contrib_drivers.drivers.ERAInstruments.erasynth.ERASynthPlusPlus>`.
+"""
+from __future__ import annotations
+
+from typing import Any, Union, Tuple
+import time
+import json
+import logging
+
+from qcodes import VisaInstrument, Parameter, validators
+
+try:
+    import pyvisa
+except ModuleNotFoundError as e:
+    raise ModuleNotFoundError(
+        "ERAInstrument drivers require the `pyvisa` package to be installed.\n"
+        "Install it with `pip install pyvisa` and try again."
+    ) from e
+
+logger = logging.getLogger(__name__)
+
+BAUDRATE = 115200
+
+_CMD_TO_JSON_MAPPING: dict = {
+    # We will treat these differently when confirming the value
+    # because the instrument replies with a sentence containing the value.
+    # which is much faster than reading the full SJON
+    # as fast as possible
+    # "P0": "rfoutput",
+    # "A": "amplitude",
+    # "F": "frequency",
+    "P1": "reference_int_ext",
+    "P5": "reference_tcxo_ocxo",
+    "MS": "modulation_on_off",
+    "M2": "modulation_signal_waveform",
+    "M1": "modulation_source",
+    "M0": "modulation_type",
+    "M3": "modulation_freq",
+    "M5": "modulation_am_depth",
+    "M4": "modulation_fm_deviation",
+    "M6": "modulation_pulse_period",
+    "M7": "modulation_pulse_width",
+    "SS": "sweep_start_stop",
+    "S0": "sweep_trigger",
+    "S1": "sweep_start",
+    "S2": "sweep_stop",
+    "S3": "sweep_step",
+    "S4": "sweep_dwell",
+    "P9": "phase_noise_mode",
+    "PEW": "wifi_mode",
+    "PES0": "wifi_sta_ssid",
+    "PEP0": "wifi_sta_password",
+    "PES1": "wifi_ap_ssid",
+    "PEP1": "wifi_ap_password",
+    "PEI": "wifi_ip_address",
+    "PEN": "wifi_subnet_address",
+    "PEG": "wifi_gateway_address",
+}
+"""
+A mapping used to read back certain commands to ensure they have been set.
+This is necessary due to non-deterministic communication times.
+"""
+
+class ERASynthBase(VisaInstrument):
+    r"""
+    A Base class for the ERASynth/ERASynth+/ERASynth++ instruments.
+
+    Example:
+
+    .. code-block::
+
+        from qcodes import Instrument
+        from qcodes_contrib_drivers.drivers.ERAInstruments import ERASynthPlus
+
+        Instrument.close_all()
+
```

Comment:
what is the reason for this? shall we leave this up to the users to close all instruments if they want/need it?
```suggestion
```